src/loki/reach.cc

```diff
@@ -0,0 +1,186 @@
+  };
+
+  // seed the expansion with a place to start expanding from
+  if (edge_filter(edge) > 0)
+    enqueue(edge->endnode());
+
+  // get outbound reach by doing a simple forward expansion until you either hit the max_reach
+  // or you can no longer expand
+  while (direction & kOutbound && queue.size() + done.size() - transitions < max_reach &&
+         !queue.empty()) {
+    // increase the reach and get the nodes id
+    auto node_id = GraphId(*done.insert(*queue.begin()).first);
+    // pop the node from the queue
+    queue.erase(queue.begin());
+    // expand from the node
+    if (!reader.GetGraphTile(node_id, tile))
+      continue;
+    for (const auto& edge : tile->GetDirectedEdges(node_id)) {
+      // if this edge is traversable we enqueue its end node
+      if (edge_filter(&edge) > 0)
+        enqueue(edge.endnode());
+    }
+  }
+  // settled nodes + will be settled nodes - duplicated transitions nodes
+  reach.outbound =
+      std::min(static_cast<uint32_t>(queue.size() + done.size() - transitions), max_reach);
+
+  // TODO: move to graphreader
+  // helper lambdas to get the begin node of an edge by using its opposing edges end node
+  auto begin_node = [&](const DirectedEdge* edge) -> GraphId {
+    // grab the node
+    if (!reader.GetGraphTile(edge->endnode(), tile))
+      return {};
+    const auto* node = tile->node(edge->endnode());
+    // grab the opp edges end node
+    const auto* opp_edge = tile->directededge(node->edge_index() + edge->opp_index());
+    return opp_edge->endnode();
+  };
+
+  // seed the expansion with a place to start expanding from
+  done.clear();
+  queue.clear();
+  transitions = 0;
+  if (edge_filter(edge) > 0)
+    enqueue(begin_node(edge));
+
+  // get inbound reach by doing a simple reverse expansion until you either hit the max_reach
+  // or you can no longer expand
+  while (direction & kInbound && queue.size() + done.size() - transitions < max_reach &&
+         !queue.empty()) {
+    // increase the reach and get the nodes id
+    auto node_id = GraphId(*done.insert(*queue.begin()).first);
+    // pop the node from the queue
+    queue.erase(queue.begin());
+    // expand from the node
+    if (!reader.GetGraphTile(node_id, tile))
+      continue;
+    for (const auto& edge : tile->GetDirectedEdges(node_id)) {
+      // get the opposing edge
+      if (!reader.GetGraphTile(edge.endnode(), tile))
+        continue;
+      const auto* node = tile->node(edge.endnode());
+      const auto* opp_edge = tile->directededge(node->edge_index() + edge.opp_index());
+      // if this opposing edge is traversable we enqueue its begin node
+      if (edge_filter(opp_edge) > 0)
+        enqueue(edge.endnode());
+    }
+  }
+  // settled nodes + will be settled nodes - duplicated transitions nodes
+  reach.inbound =
+      std::min(static_cast<uint32_t>(queue.size() + done.size() - transitions), max_reach);
+
+  return reach;
+}
+
+/*
+directed_reach Reach(const DirectedEdge* edge,
```

Comment:
this is a sketch of a more performant algorithm where we try to "share" reach between edges on the same paths (similar to the flood fill or coloring approach we had before). this approach relies on a long lived cache that edges share with each other when their paths cross/merge. its just a sketch and we can delete it if we dont want/need to do it